server.go

```diff
@@ -47,7 +62,19 @@ func (server *Server) Run() {
 			conn: conn,
 		}
 
-		go client.handleRequest()
+		// TODO move ctx from here
+		ctx := context.TODO()
+		if err := sem.Acquire(ctx, 1); err != nil {
+			slog.Error("Failed to acquire semaphore: %v", err)
+			break
+		}
+
+		go func() {
+			defer sem.Release(1)
```

Comment:
Isn't putting `defer` before `doSomething()` a GO idiomatic way of using `defer`? For example in `http` [package](https://pkg.go.dev/net/http):

```
defer resp.Body.Close()
body, err := io.ReadAll(resp.Body)
```

or in [handleRequest()](https://github.com/IBM/opentalaria/blob/38c5d6fa88a1dc1fd26b09c1fc46b0f59a45a283/server.go#L54C23-L54C38) in the same file you `defer` first. I was always thinking this is a GO way of using `defer`? 

Also `defer` not only defers the execution of a function it also close files, network connections among other things that the OS expects to close https://go.dev/blog/defer-panic-and-recover 

I am open to discussion.

Reply:
it is the idiomatic way, but I find it redundant if there are no complex flows under it, but just one function call. In the example in the blog post, using defer is the correct behaviour, otherwise you might exit the stack before closing the resources and defer makes sure to clean everything up. But here it makes no difference if you call 
```
defer sem.Release(1)
client.handleRequest()
```
or
```
client.handleRequest()
sem.Release(1)
```

I just think the second one is a bit more readable. That's purely personal preference, it doesn't make a huge difference either way, the code will behave exactly the same way.

Reply:
I think we should stick with the Go way? For someone new to Go, it will be easier to understand since most packages use `defer` in this manner. This approach also encourages them to explore and read more about `defer`. Additionally, if we have a Java developer, we want them to learn Go and write it GO way rather than writing Go code as if they were writing Java.

Reply:
agree